glean-core/rlb/src/lib.rs

```diff
@@ -119,7 +127,7 @@ where
 /// * `cfg` - the `Configuration` options to initialize with.
 /// * `client_info` - the `ClientInfoMetrics` values used to set Glean
 ///   core metrics.
-pub fn initialize(cfg: Configuration, client_info: ClientInfoMetrics) {
+pub fn initialize(cfg: Configuration<FakeUploader>, client_info: ClientInfoMetrics) {
```

Comment:
@brizental @badboy that's my other comment. See how this becomes `Configuration<FakeUploader>` instead of just `Configuration`.

I can probably make `initialize` generic, too, but then that becomes a nightmare of `"T" cannot be sent between threads safely`. Thoughts?

Reply:
Maybe if you make the generic a bit less generic? Like, `T: PingUploader + Send`? Then you would probably be able to send `T` between threads. The error message you got maybe lays out the traits it is missing.

Reply:
Yup, that would fix it (as long as I use `T: 'static +  net::PingUploader + Send` as trait bounds). However, this changes the API so that `initialize` will always require a type to be passed. It would be great to have something along the lines of what @badboy  suggests above

Reply:
Hm, IIRC there is a way that you could only allow to implement a given trait on something that already implements other traits. Then you would be able to just say `T: net::PingUploader` and that would have all the other trait dependencies implicitly? Let me check that... Indeed there is a way: https://doc.rust-lang.org/book/ch19-03-advanced-traits.html#using-supertraits-to-require-one-traits-functionality-within-another-trait I have never tried this out, heh, but worth a test.

Reply:
It still would require users to always call `initialize<SomeUploader>(cfg, client_metrics)`.

One potential way around this would be to slightly change the General API:

```rust
pub fn initialize(cfg: Configuration<net::HttpUploader>, client_info: ClientInfoMetrics) {
    initialize_with_uploader::<net::HttpUploader>(cfg, client_info);
}

pub fn initialize_with_uploader<T: 'static +  net::PingUploader + Send>(cfg: Configuration<T>, client_info: ClientInfoMetrics) {
  // ... all the init code
}
```

@brizental @badboy would this be idiomatic API design? It feels a bit awkward to me compared to the failed example from the other comment

Reply:
I don't think this is a bad choice, I am not sure it is idiomatic honestly, but it definitely is clear. I'd go with that, as I can't think of any other workarounds for this issue.

Reply:
Having two separate functions is kinda idiomatic in Rust.
However I don't think _not_ passing in the actual uploader is idiomatic. It should be up to the caller to create and allocate an uploader.